[torch][ao] Fix module name filter for pytorch2 quantization for underscores

dulinriley opened the pull request:
Summary:
There was a bug in the module name filter for modules that had an underscore
already in them, as it was replaced with a "dot" notation.
This is because it was thought that underscores always meant a module separator,
but this isn't the case for modules whose name contains an underscore.

Test Plan:
Added a unit test. Before this change, that test failed (due to applying the wrong
qscheme). Now it passes.

Differential Revision: D53502771

jerryzh168 reviewed (approved):
thanks for the fix!

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1210,6 +1210,43 @@ def forward(self, x):
                 for key in n.meta:
                     self.assertEqual(n.meta[key], weight_meta[key])
 
+    def test_xnnpack_module_name_with_underscores(self) -> None:
```
can you move this to [test_xnnpack_quantizer.py](https://github.com/pytorch/pytorch/blob/main/test/quantization/pt2e/test_xnnpack_quantizer.py)?

dulinriley replied:
Moved to that file

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1210,6 +1210,43 @@ def forward(self, x):
+            is_dynamic=False,
+            observer_or_fake_quant_ctr=observer.HistogramObserver,
+        )
+        for n in m.graph.nodes:
+            if n.op == "call_function" and n.target == torch.ops.aten.linear.default:
+                self.assertIn("quantization_annotation", n.meta)
+                annotation = n.meta["quantization_annotation"]
+                self.assertIsInstance(annotation, QuantizationAnnotation)
+                self.assertEqual(annotation.input_qspec_map[n.args[0]], correct_qspec)
+                self.assertEqual(annotation.output_qspec, correct_qspec)
```
also I think it will be better to test the prepared model inserts the correct activation_post_process (observer) module instead of testing the annotation

jerryzh168 replied:
maybe you can also simplify the test by setting `foo_bar` to be not quantized while the global setting is quantized

dulinriley replied:
Did this like you said (except the inverse, global is no quant, foo_bar is quantized)

dulinriley commented:
@pytorchbot merge

jerryzh168 commented:
please fix lint before landing

dulinriley commented:
@pytorchbot merge

dulinriley commented:
@pytorchbot merge

dulinriley committed:
[torch][ao] Fix module name filter for pytorch2 quantization for underscores (#119344)

Summary:

bypass-github-pytorch-ci-checks

There was a bug in the module name filter for modules that had an underscore
already in them, as it was replaced with a "dot" notation.
This is because it was thought that underscores always meant a module separator,
but this isn't the case for modules whose name contains an underscore.

Test Plan:
Added a unit test. Before this change, that test failed (due to applying the wrong
qscheme). Now it passes.

Reviewed By: jerryzh168

Differential Revision: D53502771
